### DISCUSSION TITLE
Cannot get the Crazyflie to select OOT contoller

### DISCUSSION THREAD

**Discussion by cameronbedard** [bitcraze/discussions#1446] [Q&A]
I am attempting to build the TinMPC OOT Controller. I have verified the Makefile and Kbuild files for TinyMPC and I built the `tinympc_controller` via the app layer with tb make-app following the official app layer documentation and double checked the Makefile and Kbuild files. The tinympc-crazyflie-firmware builds but, after loading with make cload or cfloader, I launch with cfclient, the console outputs that the PID Controller is being selected. I verified that I the Crazyflie is getting the new firmware by turning off and on the lighthousev2 in the config. Further on the cfclient, the parameters indicate with a (1) under stabilizer that PID is selected and while operating the Crazyflie, I see no evidence a different controller is running through behavior or the presence of additional logs specific to the tinympc_controller.

Is there anything I am missing?

For reference: https://github.com/RoboticExplorationLab/tinympc-crazyflie-firmware

**Comment by gemenerik** [comment]
Have you seen [this blogpost](https://www.bitcraze.io/2023/02/adding-an-estimator-or-controller/) and [this example](https://github.com/IMRCLab/crazyflie-firmware-oot)? Maybe you can start with a very basic example that basically just calls the PID controller like in the blog post and let me know if that works?

**Reply by cameronbedard** [reply]
Hello, thank you for replying!

Yes, I have looked through both before and again - just now.

I re-cloned the crazyflie-firmware main repository and re-followed the blogpost. The only edits I made to the files from blogpost and example is changing the `const uin32_t tick` to a `const stabilizerStep_t stabilizerStep`. My code also differs from the example as I am attempting to run the OOT controller as an app-layer app so I have a app-config instead of a oot-config. 

Running `make -j8` and `make cload`, the firmware flashes but again in cfclient everything points to the PID controller in the Parameters and Console tab. I also attempting to build the app through `tb make_app examples/my_controller/ -j8`

Since it is happening with both TinyMPC and a OOT build example, it could be a tool chain issue on my end? But what could be causing something like that?

**Reply by gemenerik** [reply]
Technically, the app layer is an out of tree build. You can see that in the blog post the controller is implemented through an app as well, with an app-config.

Just to be clear - did the example from the blogpost (without any modification) not work for you?
Are you running the make commands from the app directory?

**Reply by cameronbedard** [reply]
That is correct. The blogpost without making any additional changes outside of the blogpost - does not work for me.

Yes I am. `.../crazyflie-firmware/examples/my_controller$ make` followed my `make cload` does not result in running an OOT controller. It doesn't start working with the Toolbelt either.

**Reply by gemenerik** [reply]
I can confirm it does not work. It appears the OOT implementation is missing from the controller selection. https://github.com/bitcraze/crazyflie-firmware/pull/1392
